### ISSUE TITLE
integration error.

### ISSUE THREAD

**Issue by koburg** [HomeAssistant-Mods/home-assistant-miele#88]
after upgrade to latest Home assistant.

2021.12.x

i get this error in log. 

nothing is changed it has been working for very long time with no errors

I did try to reinstall in HACS no luck yet.

`Denne fejl stammer fra en brugerdefineret integration.

Logger: homeassistant.setup
Source: custom_components/miele/miele_at_home.py:28 
Integration: miele (documentation, issues) 
First occurred: 17.24.45 (1 occurrences) 
Last logged: 17.24.45

Error during setup of component miele
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/requests_oauthlib/oauth2_session.py", line 477, in request
    url, headers, data = self._client.add_token(
  File "/usr/local/lib/python3.9/site-packages/oauthlib/oauth2/rfc6749/clients/base.py", line 198, in add_token
    raise TokenExpiredError()
oauthlib.oauth2.rfc6749.errors.TokenExpiredError: (token_expired) 

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/setup.py", line 229, in _async_setup_component
    result = await task
  File "/config/custom_components/miele/__init__.py", line 150, in async_setup
    data_get_devices = await client.get_devices(lang)
  File "/config/custom_components/miele/miele_at_home.py", line 47, in get_devices
    home_devices = await self._get_devices_raw(lang)
  File "/config/custom_components/miele/miele_at_home.py", line 28, in _get_devices_raw
    devices = await self.hass.async_add_executor_job(func)
  File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.9/site-packages/requests/sessions.py", line 555, in get
    return self.request('GET', url, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/requests_oauthlib/oauth2_session.py", line 496, in request
    token = self.refresh_token(
  File "/usr/local/lib/python3.9/site-packages/requests_oauthlib/oauth2_session.py", line 446, in refresh_token
    self.token = self._client.parse_request_body_response(r.text, scope=self.scope)
  File "/usr/local/lib/python3.9/site-packages/oauthlib/oauth2/rfc6749/clients/base.py", line 429, in parse_request_body_response
    self.token = parse_token_response(body, scope=scope)
  File "/usr/local/lib/python3.9/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 425, in parse_token_response
    validate_token_parameters(params)
  File "/usr/local/lib/python3.9/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 432, in validate_token_parameters
    raise_from_error(params.get('error'), params)
  File "/usr/local/lib/python3.9/site-packages/oauthlib/oauth2/rfc6749/errors.py", line 402, in raise_from_error
    raise cls(**kwargs)
oauthlib.oauth2.rfc6749.errors.InvalidGrantError: (invalid_grant) 

`

**Comment by koburg** [comment]
still no help.

I can tell you that the new client secret and client id is now longer

secret : 32 chars
client id: 36 chars

**Comment by kloknibor** [comment]
Can you retry with the latest version, Token handling has been improved.

**Comment by koburg** [comment]
stile the same LATEST HA Version and HACS

`Denne fejl stammer fra en brugerdefineret integration.

Logger: homeassistant.setup
Source: custom_components/miele/miele_at_home.py:29 
Integration: miele (documentation, issues) 
First occurred: 10.36.48 (1 occurrences) 
Last logged: 10.36.48

Error during setup of component miele
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/requests_oauthlib/oauth2_session.py", line 483, in request
    url, headers, data = self._client.add_token(
  File "/usr/local/lib/python3.9/site-packages/oauthlib/oauth2/rfc6749/clients/base.py", line 217, in add_token
    raise TokenExpiredError()
oauthlib.oauth2.rfc6749.errors.TokenExpiredError: (token_expired) 

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/setup.py", line 235, in _async_setup_component
    result = await task
  File "/config/custom_components/miele/__init__.py", line 416, in async_setup
    data_get_devices = await client.get_devices(lang)
  File "/config/custom_components/miele/miele_at_home.py", line 48, in get_devices
    home_devices = await self._get_devices_raw(lang)
  File "/config/custom_components/miele/miele_at_home.py", line 29, in _get_devices_raw
    devices = await self.hass.async_add_executor_job(func)
  File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.9/site-packages/requests/sessions.py", line 542, in get
    return self.request('GET', url, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/requests_oauthlib/oauth2_session.py", line 502, in request
    token = self.refresh_token(
  File "/usr/local/lib/python3.9/site-packages/requests_oauthlib/oauth2_session.py", line 452, in refresh_token
    self.token = self._client.parse_request_body_response(r.text, scope=self.scope)
  File "/usr/local/lib/python3.9/site-packages/oauthlib/oauth2/rfc6749/clients/base.py", line 448, in parse_request_body_response
    self.token = parse_token_response(body, scope=scope)
  File "/usr/local/lib/python3.9/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 441, in parse_token_response
    validate_token_parameters(params)
  File "/usr/local/lib/python3.9/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 448, in validate_token_parameters
    raise_from_error(params.get('error'), params)
  File "/usr/local/lib/python3.9/site-packages/oauthlib/oauth2/rfc6749/errors.py", line 399, in raise_from_error
    raise cls(**kwargs)
oauthlib.oauth2.rfc6749.errors.InvalidGrantError: (invalid_grant) 

`

**Comment by emilhauk** [comment]
I'm also getting this error, running `HASS core-2022.4.5`, `HACS 1.24.5`, and `home-assistant-miele 0.9.3`. I have verified that the credentials are correct by performing a manual test with the api using [swagger test documentation site](https://www.miele.com/developer/swagger-ui/index.html).

I'm using the official [ghcr.io/home-assistant/home-assistant:stable](https://github.com/home-assistant/core/pkgs/container/home-assistant)
```bash
bash-5.1# cat /OFFICIAL_IMAGE 
098c6d446eaee944ffa11acf8262a1929d81cbef;refs/tags/2022.4.5;release;frenck
```

These are the oauth libraries installed in the container:
```bash
bash-5.1# pip3 list | grep oauth
google-auth-oauthlib             0.5.1
oauth2client                     4.1.3
oauthlib                         3.2.0
requests-oauth                   0.4.1
requests-oauthlib                1.3.1
```

I've never done any python debugging, but let me know if you need anything else from me.

Error log:
```text
Logger: homeassistant.setup
Source: custom_components/miele/miele_at_home.py:29
Integration: miele ([documentation](https://github.com/HomeAssistant-Mods/home-assistant-miele), [issues](https://github.com/HomeAssistant-Mods/home-assistant-miele/issues))
First occurred: 13:34:59 (1 occurrences)
Last logged: 13:34:59

Error during setup of component miele
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/requests_oauthlib/oauth2_session.py", line 483, in request
    url, headers, data = self._client.add_token(
  File "/usr/local/lib/python3.9/site-packages/oauthlib/oauth2/rfc6749/clients/base.py", line 217, in add_token
    raise TokenExpiredError()
oauthlib.oauth2.rfc6749.errors.TokenExpiredError: (token_expired) 

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/setup.py", line 235, in _async_setup_component
    result = await task
  File "/config/custom_components/miele/__init__.py", line 152, in async_setup
    data_get_devices = await client.get_devices(lang)
  File "/config/custom_components/miele/miele_at_home.py", line 48, in get_devices
    home_devices = await self._get_devices_raw(lang)
  File "/config/custom_components/miele/miele_at_home.py", line 29, in _get_devices_raw
    devices = await self.hass.async_add_executor_job(func)
  File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/local/lib/python3.9/site-packages/requests/sessions.py", line 542, in get
    return self.request('GET', url, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/requests_oauthlib/oauth2_session.py", line 502, in request
    token = self.refresh_token(
  File "/usr/local/lib/python3.9/site-packages/requests_oauthlib/oauth2_session.py", line 452, in refresh_token
    self.token = self._client.parse_request_body_response(r.text, scope=self.scope)
  File "/usr/local/lib/python3.9/site-packages/oauthlib/oauth2/rfc6749/clients/base.py", line 448, in parse_request_body_response
    self.token = parse_token_response(body, scope=scope)
  File "/usr/local/lib/python3.9/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 441, in parse_token_response
    validate_token_parameters(params)
  File "/usr/local/lib/python3.9/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 448, in validate_token_parameters
    raise_from_error(params.get('error'), params)
  File "/usr/local/lib/python3.9/site-packages/oauthlib/oauth2/rfc6749/errors.py", line 399, in raise_from_error
    raise cls(**kwargs)
oauthlib.oauth2.rfc6749.errors.InvalidGrantError: (invalid_grant)
```

**Comment by kloknibor** [comment]
Did you manually install any oath libraries? This should not be required (anymore). What happens if you remove the miele token and reboot?

**Comment by emilhauk** [comment]
First and foremost: My issue is now resolved!
I found two `.miele-token-cache`-files. Once in `config` and the other in `config/.storage`. Deleting the first one seemed not to do anything. Then I found the one in `config/.storage` (i guess this module / or hass it self relocated it at some point). Deleting that triggered a new login-process (that seems normal since I requested a new client id / secret when I began having issues). Now I have a valid token for the next month :tada:  <-- @koburg hint hint :)

I'm a bit unsure as to why this issue crept up. I have been using your addon for about a year without issues, so the refresh must have worked at some point, and I did not replace the token until I had issues.

Thanks for your help and addon @kloknibor!

Edit: As for the oauth installation. HACS readme still have that as a prerequisite when I look at it. I startet from a fresh image now. No installation required as you said, but the HACS-readme should perhaps be updated.
